File: android/app/src/main/java/com/mochi/keyboard/data/StorageRepository.kt
package com.mochi.keyboard.data

import android.net.Uri
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.storageMetadata
import kotlinx.coroutines.tasks.await
import java.util.UUID

/**
 * Upload paths/limits must match storage.rules exactly: `avatars/{uid}/{fileName}` and
 * `themes/{uid}/{fileName}`, both capped in the rules themselves (5MB avatars, 10MB theme images,
 * images only) - those limits aren't re-validated here, an oversized/wrong-type upload is rejected
 * server-side by the rules and this call throws. contentType is forced to "image/jpeg" rather than
 * left to putFile()'s automatic ContentResolver-based inference - a picked content:// Uri (e.g. from
 * the system photo picker) doesn't always resolve a type Storage's rules recognize as `image/.*`,
 * which the rules then reject as a permission error, not a type error (confirmed against the local
 * Storage emulator: an unresolved/generic contentType read back as "User does not have permission to
 * access this object", not a clearer type-mismatch message).
 */
class StorageRepository(private val storage: FirebaseStorage) {
    private val jpegMetadata = storageMetadata { contentType = "image/jpeg" }

    suspend fun uploadThemeImage(uid: String, imageUri: Uri): String {
        val ref = storage.reference.child("themes/$uid/${UUID.randomUUID()}.jpg")
        ref.putFile(imageUri, jpegMetadata).await()
        return ref.downloadUrl.await().toString()
    }

    suspend fun uploadAvatar(uid: String, imageUri: Uri): String {
        val ref = storage.reference.child("avatars/$uid/${UUID.randomUUID()}.jpg")
        ref.putFile(imageUri, jpegMetadata).await()
        return ref.downloadUrl.await().toString()
    }
}

File: ios/MochiApp/Data/StorageRepository.swift
import FirebaseStorage
import Foundation

/// Upload paths/limits must match storage.rules exactly: `avatars/{uid}/{fileName}` and
/// `themes/{uid}/{fileName}`, both capped in the rules themselves (5MB avatars, 10MB theme images,
/// images only) — those limits aren't re-validated here, an oversized/wrong-type upload is rejected
/// server-side and this call throws. contentType is forced to "image/jpeg" rather than left to
/// inference, same contract android/.../data/StorageRepository.kt uses (a system photo picker's
/// data doesn't always carry a type Storage's rules recognize as `image/.*`).
final class StorageRepository {
    private let storage: Storage

    init(storage: Storage) {
        self.storage = storage
    }

    private func upload(_ imageData: Data, to path: String) async throws -> String {
        let ref = storage.reference().child(path)
        let metadata = StorageMetadata()
        metadata.contentType = "image/jpeg"
        _ = try await ref.putDataAsync(imageData, metadata: metadata)
        return try await ref.downloadURL().absoluteString
    }

    func uploadThemeImage(uid: String, imageData: Data) async throws -> String {
        try await upload(imageData, to: "themes/\(uid)/\(UUID().uuidString).jpg")
    }

    func uploadAvatar(uid: String, imageData: Data) async throws -> String {
        try await upload(imageData, to: "avatars/\(uid)/\(UUID().uuidString).jpg")
    }
}
